src/http_server.rs

```diff
@@ -434,6 +440,124 @@ async fn get_dataset_metadata(
     axum::Json(AvailableDatasetApiResponse::new(metadata, first_block))
 }
 
+async fn get_blocknumber_by_timestamp(
+    Path((_, timestamp)): Path<(DatasetId, u64)>,
+    Extension(req): Extension<RequestId>,
+    Extension(network): Extension<Arc<NetworkClient>>,
+    Extension(task_manager): Extension<Arc<TaskManager>>,
+    Extension(config): Extension<Arc<Config>>,
+    dataset: DatasetConfig,
+) -> impl IntoResponse {
+    let ts = timestamp * 1000; // milliseconds
+    let dataset_id = dataset
+        .network_id
+        .expect("invalid dataset name should have been handled in request parser");
+
+    let Ok(chunk) = network.find_chunk_by_timestamp(&dataset_id, ts) else {
+        return RequestError::NoData.into_response();
+    };
+
+    let Ok(pquery) = build_blocknumber_query(&dataset.kind, chunk.first_block, chunk.last_block)
+    else {
+        tracing::warn!("cannot build blocknumber query for {}", dataset_id);
+        return (
+            StatusCode::NOT_FOUND,
+            format!("No data for dataset {dataset_id}"),
+        )
+            .into_response();
+    };
+
+    let mut request = build_request(
+        &config,
+        req.header_value().to_str().unwrap_or(""),
+        pquery,
+        &dataset_id,
+        &dataset.default_name,
+        1,
+    );
+
+    request.query.prepare_for_network();
+
+    let stream = match task_manager.spawn_stream(request).await {
+        Ok(stream) => stream,
+        Err(e) => {
+            tracing::warn!("spawn stream error: {:?}", e);
+            return e.into_response();
+        }
+    };
+
+    pin_mut!(stream);
+
+    let js = stream_to_string(
+        stream.map(|result| std::io::Result::Ok(tokio_util::bytes::Bytes::from_owner(result))),
+    )
+    .await;
+    if let Err(e) = js {
+        tracing::warn!("stream processing error: {:?}", e);
+        return (
+            StatusCode::NOT_FOUND,
+            format!("stream processing error: {:?}", e),
+        )
+            .into_response();
+    };
+    let js = js.unwrap();
+
+    let n = match find_block_in_chunk(timestamp, &js) {
+        Ok(n) => n,
+        Err(e) => {
+            tracing::warn!("cannot find blocknumber in chunk: {:?}", e);
+            return (StatusCode::NOT_FOUND, format!("{:?}", e)).into_response();
+        }
+    };
+
+    axum::Json(json!(n)).into_response()
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct BlockSummary {
+    header: SummaryHeader,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct SummaryHeader {
+    number: u64,
+    timestamp: u64,
+}
+
+fn find_block_in_chunk(ts: u64, js: &str) -> Result<u64, anyhow::Error> {
```

Comment:
How long does this function work? It's probably worth annotating with `#[tracing::instrument]`. If it takes longer than 100ms, it should probably be called on a blocking thread: https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html